tests/test_websocket_subscription.py

```diff
@@ -351,3 +368,40 @@ def test_websocket_subscription_sync(server, subscription_str):
         count -= 1
 
     assert count == -1
+
+
+@pytest.mark.parametrize("server", [server_countdown], indirect=True)
+@pytest.mark.parametrize("subscription_str", [countdown_subscription_str])
+def test_websocket_subscription_sync_graceful_shutdown(server, subscription_str):
```

Comment:
This one is giving me errors at my terminal:

<details>
<summary>Error Log</summary>

```console
_ test_websocket_subscription_sync_graceful_shutdown[\n    subscription {{\n      countdown (count: {count}) {{\n        number\n      }}\n    }}\n-server_countdown] _

server = <tests.conftest.WebSocketServer object at 0x0000022CAB95AF40>
subscription_str = '\n    subscription {{\n      countdown (count: {count}) {{\n        number\n      }}\n    }}\n'

    @pytest.mark.parametrize("server", [server_countdown], indirect=True)
    @pytest.mark.parametrize("subscription_str", [countdown_subscription_str])
    def test_websocket_subscription_sync_graceful_shutdown(server, subscription_str):

        url = f"ws://{server.hostname}:{server.port}/graphql"
        print(f"url = {url}")

        sample_transport = WebsocketsTransport(url=url)

        client = Client(transport=sample_transport)

        count = 10
        subscription = gql(subscription_str.format(count=count))

        with pytest.raises(KeyboardInterrupt):
            for result in client.subscribe(subscription):

                number = result["number"]
                print(f"Number received: {number}")

                assert number == count

                if count == 5:

                    # Simulate a KeyboardInterrupt in the generator
                    asyncio.ensure_future(
                        client.session._generator.athrow(KeyboardInterrupt)
                    )

                count -= 1

        assert count == 4

        # Check that the server received a connection_terminate message last
>       assert logged_messages.pop() == '{"type": "connection_terminate"}'
E       assert '{"id": "1", ...  }\\n}\\n"}}' == '{"type": "co...n_terminate"}'
E         - {"type": "connection_terminate"}
E         + {"id": "1", "type": "start", "payload": {"variables": {}, "operationName": "", "query": "subscription {\n  countdown(count: 10) {\n    number\n  }\n}\n"}}

tests\test_websocket_subscription.py:407: AssertionError
----------------------------------------------------------------------- Captured stdout setup ------------------------------------------------------------------------
Starting server
Server started on port 60263
------------------------------------------------------------------------ Captured stdout call ------------------------------------------------------------------------
url = ws://127.0.0.1:60263/graphql

Countdown started from: 10
Number received: 10
Number received: 9
Number received: 8
Number received: 7
Number received: 6
Number received: 5
Now stopping task is cancelled
Now keepalive task is cancelled
------------------------------------------------------------------------ Captured stderr call ------------------------------------------------------------------------
server < GET /graphql HTTP/1.1
DEBUG:websockets.server:server < GET /graphql HTTP/1.1
server < Headers([('Host', '127.0.0.1:60263'), ('Upgrade', 'websocket'), ('Connection', 'Upgrade'), ('Sec-WebSocket-Key', 'nJ4ZqrH1lHG5kjDJeQ+dxw=='), ('Sec-WebSocket-Version', '13'), ('Sec-WebSocket-Extensions', 'permessage-deflate; client_max_window_bits'), ('Sec-WebSocket-Protocol', 'graphql-ws'), ('User-Agent', 'Python/3.8 websockets/8.1')])
DEBUG:websockets.server:server < Headers([('Host', '127.0.0.1:60263'), ('Upgrade', 'websocket'), ('Connection', 'Upgrade'), ('Sec-WebSocket-Key', 'nJ4ZqrH1lHG5kjDJeQ+dxw=='), ('Sec-WebSocket-Version', '13'), ('Sec-WebSocket-Extensions', 'permessage-deflate; client_max_window_bits'), ('Sec-WebSocket-Protocol', 'graphql-ws'), ('User-Agent', 'Python/3.8 websockets/8.1')])
server > HTTP/1.1 101 Switching Protocols
DEBUG:websockets.server:server > HTTP/1.1 101 Switching Protocols
server > Headers([('Upgrade', 'websocket'), ('Connection', 'Upgrade'), ('Sec-WebSocket-Accept', 'G6QrCXZ9JtFkyB9j/NdkzpREsOs='), ('Sec-WebSocket-Extensions', 'permessage-deflate'), ('Date', 'Sat, 20 Jun 2020 02:56:30 GMT'), ('Server', 'Python/3.8 websockets/8.1')])
DEBUG:websockets.server:server > Headers([('Upgrade', 'websocket'), ('Connection', 'Upgrade'), ('Sec-WebSocket-Accept', 'G6QrCXZ9JtFkyB9j/NdkzpREsOs='), ('Sec-WebSocket-Extensions', 'permessage-deflate'), ('Date', 'Sat, 20 Jun 2020 02:56:30 GMT'), ('Server', 'Python/3.8 websockets/8.1')])
>>> {"type": "connection_init", "payload": {}}
INFO:gql.transport.websockets:>>> {"type": "connection_init", "payload": {}}
<<< {"type":"connection_ack"}
INFO:gql.transport.websockets:<<< {"type":"connection_ack"}
>>> {"id": "1", "type": "start", "payload": {"variables": {}, "operationName": "", "query": "subscription {\n  countdown(count: 10) {\n    number\n  }\n}\n"}}
INFO:gql.transport.websockets:>>> {"id": "1", "type": "start", "payload": {"variables": {}, "operationName": "", "query": "subscription {\n  countdown(count: 10) {\n    number\n  }\n}\n"}}
<<< {"type":"ka"}
INFO:gql.transport.websockets:<<< {"type":"ka"}
<<< {"type":"data","id":"1","payload":{"data":{"number":10}}}
INFO:gql.transport.websockets:<<< {"type":"data","id":"1","payload":{"data":{"number":10}}}
<<< {"type":"data","id":"1","payload":{"data":{"number":9}}}
INFO:gql.transport.websockets:<<< {"type":"data","id":"1","payload":{"data":{"number":9}}}
<<< {"type":"ka"}
INFO:gql.transport.websockets:<<< {"type":"ka"}
<<< {"type":"data","id":"1","payload":{"data":{"number":8}}}
INFO:gql.transport.websockets:<<< {"type":"data","id":"1","payload":{"data":{"number":8}}}
<<< {"type":"ka"}
INFO:gql.transport.websockets:<<< {"type":"ka"}
<<< {"type":"data","id":"1","payload":{"data":{"number":7}}}
INFO:gql.transport.websockets:<<< {"type":"data","id":"1","payload":{"data":{"number":7}}}
<<< {"type":"ka"}
INFO:gql.transport.websockets:<<< {"type":"ka"}
<<< {"type":"data","id":"1","payload":{"data":{"number":6}}}
INFO:gql.transport.websockets:<<< {"type":"data","id":"1","payload":{"data":{"number":6}}}
<<< {"type":"ka"}
INFO:gql.transport.websockets:<<< {"type":"ka"}
<<< {"type":"data","id":"1","payload":{"data":{"number":5}}}
INFO:gql.transport.websockets:<<< {"type":"data","id":"1","payload":{"data":{"number":5}}}
<<< {"type":"ka"}
INFO:gql.transport.websockets:<<< {"type":"ka"}
<<< {"type":"data","id":"1","payload":{"data":{"number":4}}}
INFO:gql.transport.websockets:<<< {"type":"data","id":"1","payload":{"data":{"number":4}}}
<<< {"type":"ka"}
INFO:gql.transport.websockets:<<< {"type":"ka"}
<<< {"type":"data","id":"1","payload":{"data":{"number":3}}}
INFO:gql.transport.websockets:<<< {"type":"data","id":"1","payload":{"data":{"number":3}}}
<<< {"type":"ka"}
INFO:gql.transport.websockets:<<< {"type":"ka"}
<<< {"type":"data","id":"1","payload":{"data":{"number":2}}}
INFO:gql.transport.websockets:<<< {"type":"data","id":"1","payload":{"data":{"number":2}}}
<<< {"type":"ka"}
INFO:gql.transport.websockets:<<< {"type":"ka"}
>>> {"type": "connection_terminate"}
INFO:gql.transport.websockets:>>> {"type": "connection_terminate"}
<<< {"type":"data","id":"1","payload":{"data":{"number":1}}}
INFO:gql.transport.websockets:<<< {"type":"data","id":"1","payload":{"data":{"number":1}}}
<<< {"type":"ka"}
INFO:gql.transport.websockets:<<< {"type":"ka"}
<<< {"type":"data","id":"1","payload":{"data":{"number":0}}}
INFO:gql.transport.websockets:<<< {"type":"data","id":"1","payload":{"data":{"number":0}}}
<<< {"type":"ka"}
INFO:gql.transport.websockets:<<< {"type":"ka"}
<<< {"type":"ka"}
INFO:gql.transport.websockets:<<< {"type":"ka"}
Exiting _receive_data_loop()
DEBUG:gql.transport.websockets:Exiting _receive_data_loop()
------------------------------------------------------------------------- Captured log call --------------------------------------------------------------------------
DEBUG    websockets.server:server.py:237 server < GET /graphql HTTP/1.1
DEBUG    websockets.server:server.py:238 server < Headers([('Host', '127.0.0.1:60263'), ('Upgrade', 'websocket'), ('Connection', 'Upgrade'), ('Sec-WebSocket-Key', 'nJ4ZqrH1lHG5kjDJeQ+dxw=='), ('Sec-WebSocket-Version', '13'), ('Sec-WebSocket-Extensions', 'permessage-deflate; client_max_window_bits'), ('Sec-WebSocket-Protocol', 'graphql-ws'), ('User-Agent', 'Python/3.8 websockets/8.1')])
DEBUG    websockets.server:server.py:256 server > HTTP/1.1 101 Switching Protocols
DEBUG    websockets.server:server.py:257 server > Headers([('Upgrade', 'websocket'), ('Connection', 'Upgrade'), ('Sec-WebSocket-Accept', 'G6QrCXZ9JtFkyB9j/NdkzpREsOs='), ('Sec-WebSocket-Extensions', 'permessage-deflate'), ('Date', 'Sat, 20 Jun 2020 02:56:30 GMT'), ('Server', 'Python/3.8 websockets/8.1')])
INFO     gql.transport.websockets:websockets.py:150 >>> {"type": "connection_init", "payload": {}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"connection_ack"}
INFO     gql.transport.websockets:websockets.py:150 >>> {"id": "1", "type": "start", "payload": {"variables": {}, "operationName": "", "query": "subscription {\n  countdown(count: 10) {\n    number\n  }\n}\n"}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"ka"}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"data","id":"1","payload":{"data":{"number":10}}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"data","id":"1","payload":{"data":{"number":9}}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"ka"}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"data","id":"1","payload":{"data":{"number":8}}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"ka"}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"data","id":"1","payload":{"data":{"number":7}}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"ka"}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"data","id":"1","payload":{"data":{"number":6}}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"ka"}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"data","id":"1","payload":{"data":{"number":5}}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"ka"}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"data","id":"1","payload":{"data":{"number":4}}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"ka"}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"data","id":"1","payload":{"data":{"number":3}}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"ka"}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"data","id":"1","payload":{"data":{"number":2}}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"ka"}
INFO     gql.transport.websockets:websockets.py:150 >>> {"type": "connection_terminate"}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"data","id":"1","payload":{"data":{"number":1}}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"ka"}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"data","id":"1","payload":{"data":{"number":0}}}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"ka"}
INFO     gql.transport.websockets:websockets.py:172 <<< {"type":"ka"}
DEBUG    gql.transport.websockets:websockets.py:362 Exiting _receive_data_loop()
---------------------------------------------------------------------- Captured stdout teardown ----------------------------------------------------------------------
Stopping server
Server stopped



---------------------------------------------------------------------- Captured stderr teardown ----------------------------------------------------------------------
ERROR:asyncio:Task exception was never retrieved
future: <Task finished name='Task-1525' coro=<<async_generator_athrow without __name__>()> exception=KeyboardInterrupt()>
Traceback (most recent call last):
  File "D:\Programming\py_projects\graphql-python-repos\gql\tests\test_websocket_subscription.py", line 388, in test_websocket_subscription_sync_graceful_shutdown
    for result in client.subscribe(subscription):
  File "D:\Programming\py_projects\graphql-python-repos\gql\gql\client.py", line 153, in subscribe
    result = loop.run_until_complete(generator_task)
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\base_events.py", line 599, in run_until_complete
    self.run_forever()
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\windows_events.py", line 316, in run_forever
    super().run_forever()
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\base_events.py", line 567, in run_forever
    self._run_once()
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\base_events.py", line 1855, in _run_once
    handle._run()
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\events.py", line 81, in _run
    self._context.run(self._callback, *self._args)
  File "D:\Programming\py_projects\graphql-python-repos\gql\gql\transport\websockets.py", line 401, in subscribe
    yield execution_result
KeyboardInterrupt
ERROR:asyncio:Error on reading from the event loop self pipe
loop: <ProactorEventLoop running=True closed=False debug=False>
Traceback (most recent call last):
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\windows_events.py", line 453, in finish_recv
    return ov.getresult()
OSError: [WinError 995] La operación de E/S se anuló por una salida de subproceso o por una solicitud de aplicación

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\proactor_events.py", line 768, in _loop_self_reading
    f.result()  # may raise
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\windows_events.py", line 808, in _poll
    value = callback(transferred, key, ov)
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\windows_events.py", line 457, in finish_recv
    raise ConnectionResetError(*exc.args)
ConnectionResetError: [WinError 995] La operación de E/S se anuló por una salida de subproceso o por una solicitud de aplicación
----------------------------------------------------------------------- Captured log teardown ------------------------------------------------------------------------
ERROR    asyncio:base_events.py:1703 Error on reading from the event loop self pipe
loop: <ProactorEventLoop running=True closed=False debug=False>
Traceback (most recent call last):
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\windows_events.py", line 453, in finish_recv
    return ov.getresult()
OSError: [WinError 995] La operación de E/S se anuló por una salida de subproceso o por una solicitud de aplicación

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\proactor_events.py", line 768, in _loop_self_reading
    f.result()  # may raise
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\windows_events.py", line 808, in _poll
    value = callback(transferred, key, ov)
  File "d:\anaconda3\envs\gql-dev\lib\asyncio\windows_events.py", line 457, in finish_recv
    raise ConnectionResetError(*exc.args)
ConnectionResetError: [WinError 995] La operación de E/S se anuló por una salida de subproceso o por una solicitud de aplicación
========================================================================== warnings summary ==========================================================================
tests/test_aiohttp.py::test_aiohttp_query
tests/test_aiohttp.py::test_aiohttp_error_code_500
tests/test_aiohttp.py::test_aiohttp_error_code
tests/test_aiohttp.py::test_aiohttp_invalid_protocol[{}]
tests/test_aiohttp.py::test_aiohttp_invalid_protocol[qlsjfqsdlkj]
tests/test_aiohttp.py::test_aiohttp_invalid_protocol[{"not_data_or_errors": 35}]
tests/test_aiohttp.py::test_aiohttp_extra_args
tests/test_aiohttp.py::test_aiohttp_query_variable_values
  d:\anaconda3\envs\gql-dev\lib\site-packages\aiohttp\connector.py:964: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
    hosts = await asyncio.shield(self._resolve_host(

tests/test_aiohttp.py::test_aiohttp_error_code_500
tests/test_aiohttp.py::test_aiohttp_subscribe_not_supported
tests/test_aiohttp.py::test_aiohttp_cannot_connect_twice
tests/test_aiohttp.py::test_aiohttp_cannot_execute_if_not_connected
tests/test_requests.py::test_requests_error_code_500
tests/test_requests.py::test_requests_cannot_connect_twice
tests/test_requests.py::test_requests_cannot_execute_if_not_connected
  d:\anaconda3\envs\gql-dev\lib\site-packages\aiohttp\web_server.py:53: DeprecationWarning: The loop argument is deprecated since Python 3.8, and scheduled for removal in Python 3.10.
    await asyncio.gather(*coros, loop=self._loop)

-- Docs: https://docs.pytest.org/en/latest/warnings.html

----------- coverage: platform win32, python 3.8.1-final-0 -----------
Name                               Stmts   Miss  Cover   Missing
----------------------------------------------------------------
gql\__init__.py                        3      0   100%
gql\client.py                        136      0   100%
gql\dsl.py                            93      1    99%   86
gql\gql.py                             4      0   100%
gql\transport\__init__.py              3      0   100%
gql\transport\aiohttp.py              57      0   100%
gql\transport\async_transport.py      12      0   100%
gql\transport\exceptions.py           10      0   100%
gql\transport\local_schema.py         27      3    89%   42-43, 63
gql\transport\requests.py             57      0   100%
gql\transport\transport.py             8      1    88%   23
gql\transport\websockets.py          230      0   100%
gql\utils.py                           3      0   100%
----------------------------------------------------------------
TOTAL                                643      5    99%

====================================================================== short test summary info =======================================================================
FAILED tests/test_websocket_subscription.py::test_websocket_subscription_sync_graceful_shutdown[\n    subscription {{\n      countdown (count: {count}) {{\n        number\n      }}\n    }}\n-server_countdown]
======================================================= 1 failed, 137 passed, 24 skipped, 15 warnings in 7.06s =======================================================
```
</details>

Using Windows 10 with Anaconda.

Reply:
Thank you for your test on Windows.

I have reproduced this behavior on Windows 10 with python 3.8

I don't know exactly why this is happening.
Apparently the control-C signal and KeyboardInterrupt exception with asyncio is managed differently on Linux and on Windows. This might be a bug in asyncio, I'll investigate further.

Nevertheless I tested the code of this pull request manually (typing control-C during a sync subscription and checking the result) and this is working correctly. So the problem is only with the test.

I tried to use different ways to trigger the KeyboardInterrupt (like starting a new process which will kill the current one) but it is not working because the server is killed at the same time in the tests. I could try to start the server in another process but this is getting complicated.

I think that we could just skip this test if we are on Windows. Would that be ok for you ?

Reply:
As long as the real behaviour works on Windows and only fails due to the test setup, I am fine with skipping this test for Windows only. Everything else on the PR is fine.

Reply:
Ok, I'll do that then.